crypto/params.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+
+static const OSSL_PARAM *param_locate(const OSSL_PARAM *p, const char *key)
+{
+    if (p != NULL && key != NULL)
+        for (; p->key != NULL; p++)
+            if (strcmp(key, p->key) == 0)
+                return p;
+    return NULL;
+}
+
+#define PARAM_INT_GET(to, from) \
+    case sizeof(from): \
+        if (sizeof(to) >= sizeof(from)) { \
+            *val = (to)*(const from *)p->buffer; \
+            return 1; \
+        } \
+        break
+
+#define PARAM_INT_SET(to, from) \
+    case sizeof(to): \
+        if (sizeof(to) >= sizeof(from)) { \
+            *(to *)p->buffer = (to)val; \
+            return 1; \
+        } \
+        break
+
+#define PARAM_INT(name, type, pre) \
+    int OSSL_PARAM_get_##name(const OSSL_PARAM *p, const char *key, type *val) \
+    { \
+        if (val == NULL \
+            || (p = param_locate(p, key)) == NULL \
+            || (p->data_type != OSSL_PARAM_INTEGER \
+                && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER)) \
+            return 0; \
+        switch (p->buffer_size) { \
+        PARAM_INT_GET(type, pre ## 32_t); \
+        PARAM_INT_GET(type, pre ## 64_t); \
+        } \
+        return 0; \
+    } \
+    int OSSL_PARAM_set_##name(const OSSL_PARAM *p, const char *key, type val) \
+    { \
+        if ((p = param_locate(p, key)) == NULL \
+            || (p->data_type != OSSL_PARAM_INTEGER \
+                && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER)) \
+            return 0; \
+        if (p->return_size != NULL) \
+            *p->return_size = sizeof(type); \
+        switch (p->buffer_size) { \
+        PARAM_INT_SET(pre ## 32_t, type); \
+        PARAM_INT_SET(pre ## 64_t, type); \
+        } \
+        return 0; \
+    }
+
+PARAM_INT(int, int, int)
+PARAM_INT(uint, unsigned int, uint)
+PARAM_INT(long, long int, int)
+PARAM_INT(ulong, unsigned long int, uint)
+PARAM_INT(int32, int32_t, int)
+PARAM_INT(uint32, uint32_t, uint)
+PARAM_INT(int64, int64_t, int)
+PARAM_INT(uint64, uint64_t, uint)
+PARAM_INT(size_t, size_t, uint)
+
+int OSSL_PARAM_get_BN(const OSSL_PARAM *p, const char *key, BIGNUM **val)
+{
+    BIGNUM *b;
+
+    if (val == NULL || (p = param_locate(p, key)) == NULL)
+        return 0;
+
+    if (p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) {
+        b = BN_native2bn(p->buffer, (int)p->buffer_size, *val);
+        if (b != NULL) {
+            *val = b;
+            if (p->return_size != NULL)
+                *p->return_size = BN_num_bytes(b);
+            return 1;
+        }
```

Comment:
This doesn't seem quite right.  When you *get*  a BIGNUM out of an `OSSL_PARAM`, there's no reason to *set* `*p->return_size`.  However, if might be reasonable to *use* `*p->return_size`, if this function gets called by the owner of the `OSSL_PARAM`.

So, something like this?

``` C
    if (p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) {
        b = BN_native2bn(p->buffer, p->return_size ? (int)*p->return_size : (int)p->buffer_size, *val);
        if (b != NULL) {
            *val = b;
            return 1;
        }

Reply:
It's possible that the getter macros for the diverse C types should have something similar...  that's probably less important, since the size should be known beforehand, but purity might still be a thing...

Reply:
I'm not convinced this needs changing.  I am quite possibly missing something.  I'm expecting the usual calling convention to be:

```
BIGNUM *val = NULL;

OSSL_PARAM_get_BN(params, "xyzzy", &val)
```

Reply:
Ok, so what about this?

``` C
    unsigned char binbignum[4096];
    size_t bignum_bytes = 0;
    const OSSL_PARAM foo[] = {
        OSSL_PARAM_SIZED_BN("foo", &binbignum, sizeof(binbignum), &bignum_bytes),
        OSSL_PARAM_END
    };

    if (whatever_get_params(someobj, foo)) {
        BIGNUM *bn = NULL;

        if (OSSL_PARAM_get_BN(foo, "foo", &bn))
            ...
```

The relevant question is, who do you expect to call `OSSL_PARAM_get_BN`?

Reply:
Either way, *setting* `*p->return_size` in this function is just plain wrong.

Reply:
Agreed and fixed.

Reply:
Actually, my wasn't very good.  I see no reason whythe `OSSL_PARAM` ownder should call `OSSL_PARAM_get_BN` when all they have to do is `bn = BN_native2bn(binbignum, bignum_bytes, NULL);`

Still, this discussion brought clarity, so not a total loss...